Flame graph colors

brisad opened the pull request:
This adds three base colors to the flame graph, one for C++, one for JS, and one for Unknown(?). I tried to account for color blindness, but I don't know if it was successful or not. The colors will get darker the more self time a function has.

The tooltips have been extended to show the implementation (together with a color swatch), running time and self time.

Addresses #726

brisad commented on src/components/flame-graph/Canvas.js:
```diff
@@ -58,6 +58,38 @@ class FlameGraphCanvas extends React.PureComponent<Props> {
+      a = 10;
+      b = 10;
+      limit = 40;
+    }
```
So these three base colors were chosen from paletton.com. I tried to choose three colors that would account for the most common forms of color blindness, but I have no idea whether I was successful there.

I have empirically found values for the factors `a` and `b`, and `limit` by visual inspection (and some initial curve fitting), and trying to subjectively determine what looks nice. I started with a linear formula, but using a logarithm appeared to me to give a better range of colors for the gradient.

gregtatum replied:
I think this observation would be great as a code comment.

brisad commented on src/profile-logic/transforms.js:
```diff
@@ -1333,6 +1337,18 @@ export function restoreAllFunctionsInCallNodePath(
+  } else if (FUNC_MATCHES.js(thread, funcIndex)) {
+    return 'js';
+  }
+  return 'unknown';
```
I didn't know what this case implies, hence 'unknown'. But perhaps we know what this is?

gregtatum replied:
These are probably JIT code, that is generated by the JS engine, but could also be unsymbolicated native code, e.g. from a third party library like Flash.

brisad commented:
Example screenshot with visible tooltip:

![flame-graph](https://user-images.githubusercontent.com/1911402/36616012-906f9fda-18e2-11e8-823b-9228876996fe.png)

gregtatum commented:
Here is an example profile that shows how much more color does for understanding a profile: https://perfht.ml/2FaZADN

<img width="1552" alt="image" src="https://user-images.githubusercontent.com/1588648/36868394-ce9411c2-1d5d-11e8-84a8-9e9dcacd9a23.png">

<img width="1552" alt="image" src="https://user-images.githubusercontent.com/1588648/36868386-cb92ca18-1d5d-11e8-981b-452fdc939395.png">

gregtatum commented:
I feel like I could bikeshed on the colors, but the only thing I will ask on tweaking is to align them with the closest value on the [photon colors design system](https://design.firefox.com/photon/visuals/color.html). This will work for now.

Also, I think the darkest red is not accessible enough, as the black text is hard to read.

gregtatum commented on src/components/flame-graph/Canvas.js:
```diff
@@ -58,6 +58,38 @@ class FlameGraphCanvas extends React.PureComponent<Props> {
     (this: any)._hitTest = this._hitTest.bind(this);
   }
 
+  getColor(implementation: Implementation, selfTimeRelative: number): string {
```
Could you please unit test this function? I think you can break it out into its own function, as it's a pure function. You can still keep it in this file, I don't think you need to break it out into utils or anything.

gregtatum commented on src/components/flame-graph/Canvas.js:
```diff
@@ -58,6 +58,38 @@ class FlameGraphCanvas extends React.PureComponent<Props> {
+    // the lower limit of the lightness.
+    let a, b, limit;
+
+    if (implementation === 'cpp') {
```
I think here it would be nice to go through the list exhaustively using a switch statement. This way flow throws an error if the type ever changes:

```js
    switch (implementation) {
      case 'cpp':
        [h, s, l] = [358, 98, 82];
        a = 4;
        b = 5;
        limit = 55;
        break;
      case 'js':
        [h, s, l] = [188, 53, 71];
        a = 8;
        b = 6;
        limit = 40;
        break;
      case 'unknown':
        // Unknown implementation
        [h, s, l] = [77, 70, 71];
        a = 10;
        b = 10;
        limit = 40;
        break;
      default:
        throw new Error(
          `Unknown implementation case "${(implementation: empty)}".`
        );
    }
```

gregtatum commented on src/profile-logic/flame-graph.js:
```diff
@@ -14,6 +14,8 @@ export type IndexIntoFlameGraphTiming = number;
 export type FlameGraphTiming = Array<{
   start: UnitIntervalOfProfileRange[],
   end: UnitIntervalOfProfileRange[],
+  selfTimeRelative: Array<number>,
```
Could you add explanatory code comments on these new fields?

gregtatum commented on src/test/components/__snapshots__/FlameGraph.test.js.snap:
```diff
@@ -132,7 +132,7 @@ Array [
   Array [
     "set fillStyle",
-    "#ededf0",
+    "hsl(358, 98%, 82%)",
```
I really like snapshot test :)

gregtatum commented on src/types/profile-derived.js:
```diff
@@ -103,3 +105,5 @@ export type MarkerTiming = {
 };
 export type MarkerTimingRows = Array<MarkerTiming>;
+
+export type Implementation = 'js' | 'cpp' | 'unknown';
```
So at this point we have a bit of ambiguity around `Implementation`. I was chatting with Markus about this, and I think we should do the following instead:

```js
export type StackType = 'js' | 'native' | 'unsymbolicated';
```

Then in the tooltip, instead of "Implementation:" it would say "Stack Type:". The human readable labels would be:

"JavaScript"
"Native"
"Unsymbolicated Native"

I think later we will also include a PseudoStack type of entry.

brisad commented:
Thanks for the review!

I have addressed all your comments, and this PR is ready for re-review.

Tweaking the colors to be inline with the Photon colors turned out to not be so straightforward, so I had to make some bigger changes there. But I think it turned out to be way better in terms of understanding what's happening. I did add a new dependency to `d3-interpolate` though.
I wrote some more documentation which I hope is makes sense, let me know otherwise.

brisad commented:
![screenshot-2018-3-7 perf html](https://user-images.githubusercontent.com/1911402/37119786-d5095ab2-2258-11e8-92f8-0ecefb7cec2b.png)

brisad commented on src/components/flame-graph/Canvas.js:
```diff
@@ -48,6 +52,64 @@ const ROW_HEIGHT = 16;
+  switch (stackType) {
+    case 'native':
+      backgroundRange = [colors.BLUE_40, colors.BLUE_70];
+      foreground = t > 0.25 ? '#ffffff' : colors.BLUE_90;
```
I should've written this as a comment, but this flips the foreground color to white when the interpolated color reaches a color very similar to BLUE_50, since that happens when we're on about one fourth of the range. And that color should have a white foreground color according to the Photon design system.

This probably violates the `AA Large` contrast requirement for that color, but I thought this might perhaps be good enough. Having a gradient color makes it difficult (for me) to know what foreground colors are acceptable.

brisad commented on src/components/flame-graph/Canvas.js:
```diff
@@ -48,6 +52,64 @@ const ROW_HEIGHT = 16;
+    default:
+      throw new Error(`Unknown stack type case "${(stackType: empty)}".`);
+  }
+  return { background: interpolateRgb(...backgroundRange)(t), foreground };
```
D3 also has `interpolateRgbBasis` (which allows interpolation between several points, like a list of `[ORANGE_50, ORANGE_60, ORANGE_70]`) and `interpolateHsl`, but `interpolateRgb` seemed to be (after making a few informal measurements) the fastest, and provided more or less the same color ranges.

By the way, `getColors` seems to be responsible of 20% of the time in `_drawCanvas`. Not optimal, but perhaps alright for a first version? Most of it seemingly having to do with converting colors to RGB within in D3.

gregtatum replied:
Oh, and one other comment, if you can split out the foreground and background colors into separate functions, that'll save returning an object. I'm not sure the actual impact of JITed code returning objects that immediately get destructured, it may optimize that allocation out (no idea really.) I think this type of thing is important due to the perf-sensitive nature of the function.

gregtatum commented on src/components/flame-graph/Canvas.js:
```diff
@@ -48,6 +50,64 @@ const ROW_HEIGHT = 16;
 const TEXT_OFFSET_START = 3;
 const TEXT_OFFSET_TOP = 11;
 
+/* Return the background and foreground colors used to paint a box in
```
This is a small style tweak, but generally we've done the following style for long-form comments:

```
/**
 * Long form comment
 */
```

Also, thanks for writing long form comments!

brisad commented:
Thanks!
I've updated the PR to use the colors you suggested, looks much better!
I also implemented your performance enhancing suggestions, including your interpolation code. It made a huge difference!

It's great fun learning about how I can improve performance in hot paths :-D

brisad commented on src/components/flame-graph/Canvas.js:
```diff
@@ -48,6 +49,138 @@ const ROW_HEIGHT = 16;
+  stackType: StackType,
+  selfTimeRelative: number
+): string {
+  const t = Math.log(5000 * selfTimeRelative + 1) / Math.log(5001);
```
How expensive are function calls? I skipped extracting this formula into its own function.

gregtatum replied:
In unoptimized code, there is a small cost to call out to other functions, but optimized functions will often inline functions and remove that cost. The risk on using lots of functions, is that if the call uses differently shaped arguments, it can be de-optimized. This starts to get into the realm of micro-optimizations though. The key is to profile and see what is going on. The stack chart will show if the code was optimized or not.

brisad replied:
Alright, thanks for the answer.

gregtatum commented:
Looks like there are conflicts on merging.

brisad commented:
Are there still conflicts? Everything is green here on GitHub and it seems like you successfully merged master into the branch.

gregtatum commented:
@brisad I'm seeing this message:
<img width="702" alt="image" src="https://user-images.githubusercontent.com/1588648/37433934-dee9387e-27ab-11e8-9481-9f6fdad28993.png">

brisad commented:
@gregtatum Ah I see. I also see that message if I choose "Rebase and merge" from the dropdown. And that is not surprising since I merged in the master branch a couple of time into the feature branch. Using merge commits should work though. Are we always using rebase when merging into master?

Either way I'll try to get rid of the merge commits so that the branch can be cleanly rebased on master.

brisad commented:
Turned out to be pretty straightforward to resolve the conflicts locally. I've pushed a new version with everything rebased to the latest master.

gregtatum commented:
Thanks, I just merged it all in!
